client/network/src/debug_info.rs

```diff
@@ -205,26 +205,15 @@ impl NetworkBehaviour for DebugInfoBehaviour {
 		}
 	}
 
-	fn inject_node_event(
+	fn inject_event(
 		&mut self,
 		peer_id: PeerId,
+		connection: ConnectionId,
 		event: <<Self::ProtocolsHandler as IntoProtocolsHandler>::Handler as ProtocolsHandler>::OutEvent
 	) {
 		match event {
-			EitherOutput::First(event) => self.ping.inject_node_event(peer_id, event),
-			EitherOutput::Second(event) => self.identify.inject_node_event(peer_id, event),
-		}
-	}
-
-	fn inject_replaced(&mut self, peer_id: PeerId, closed_endpoint: ConnectedPoint, new_endpoint: ConnectedPoint) {
-		self.ping.inject_replaced(peer_id.clone(), closed_endpoint.clone(), new_endpoint.clone());
-		self.identify.inject_replaced(peer_id.clone(), closed_endpoint, new_endpoint.clone());
-
-		if let Some(entry) = self.nodes_info.get_mut(&peer_id) {
-			entry.endpoint = new_endpoint;
```

Comment:
The fact that this line disappeared made me realize that the `NetworkBehaviour` has no way to know if the `ConnectedPoint` passed with `inject_connected` is still in use.

Reply:
The right thing to do theoretically speaking would be to remove the `ConnectedPoint` parameter from `inject_connected` entirely. Unfortunately the `GenericProto` behaviour uses this value at the moment.

Reply:
> The fact that this line disappeared made me realize that the `NetworkBehaviour` has no way to know if the `ConnectedPoint` passed with `inject_connected` is still in use.

What do you mean by "is still in use"? By whom? Do you mean that when `inject_connected` is called, the endpoint might actually concurrently be closed by the background task e.g. on error and `inject_disconnected` may not even be called if there happens to be another connection? So `inject_connected` and `inject_disconnected` can actually be called with different endpoints? That is an interesting observation, though I'm not sure what problems that may cause, if any (since `inject_disconnected` means there is _no_ connection remaining, it obviously implies that endpoint originally passed to `inject_connected` is also closed, but if only the endpoint originally passed to `inject_connected` closes, with another connection remaining, the behaviour will indeed not know about that). The reason `inject_connected` and `inject_disconnected` are only called for the first and last connection, respectively, is of course intentional for backward compatibility. I thought we would later add new callbacks that are called for every connection, e.g. `inject_connection_established` and `inject_connection_closed`. Then we could move the `ConnectedPoint` parameter to these methods, where it is assured that the calls are always paired up for the same connected point and thus if an implementation of `NetworkBehaviour` wants to use these, it must use these callbacks.

Reply:
The consequence for `debug_info.rs` is that we're tracking, for each connection, its corresponding `ConnectedPoint`. This tracking is used only for debugging purposes, so it's not a big deal.

However, in `light_client_handler.rs` for example, this is more problematic. This module is responsible for requesting information about blocks/justifications/etc. from the network.
To give you a summary:

- When `LightClientHandler::request` is called, the request is added to a list of pending requests.
- In `poll()`, we then send this request to the `ProtocolsHandler` of a specific peer which we know we're connected to. This is done by emitting a `SendEvent`.
- If everything goes well, the `ProtocolsHandler` will later send back an event containing the response.
- However if `inject_disconnected()` is called with a peer we have sent a request to, then we have to add back the request to the list of pending requests.

Now that we have multiple connections per peer, it possible for the `ProtocolsHandler` to be silently dropped, in which case the `NetworkBehaviour` has no way to know that it has to try again the request with a different peer.

> I thought we would later add new callbacks that are called for every connection, e.g. inject_connection_established and inject_connection_closed.

I don't think there's a way to solve that problem without tackling this.

Reply:
> Now that we have multiple connections per peer, it possible for the `ProtocolsHandler` to be silently dropped, in which case the request is silently lost and the `NetworkBehaviour` has no way to know that.

That generally shouldn't be a problem though, because any such code that expects a reply to some request necessarily needs a timeout (+ bounded retry) mechanism, as seems to be the case for the [`LightClientHandler`](https://github.com/paritytech/substrate/blob/master/client/network/src/protocol/light_client_handler.rs#L862-L887). Requests can always fail by not getting a (timely) response. And any code that sends something without expecting a reply is by definition accepting potential delivery failure of the request.

Reply:
You're essentially suggesting that, even when we *know* that a request has failed, we instead just wait for the timeout.

Reply:
> You're essentially suggesting that, even when we _know_ that a request has failed, we instead just wait for the timeout.

It is not a suggestion, just an observation that I don't think this scenario should cause any problems. It is ultimately of course desirable to be aware of failed requests as soon as possible, and that can be done once the behaviour has callbacks for each connection open/close.

Reply:
Since `libp2p-0.17` included the new callbacks, I addressed this concern in https://github.com/paritytech/substrate/pull/5278/commits/31cdf4ff6465dcf4be743341d092616637b60f06. TL;DR: The `light_client_handler` tracks the connections a request has been sent with and re-enqueues any outstanding requests on a connection as pending requests when that connection closes.